distributed/shuffle/_shuffle_extension.py

```diff
@@ -588,29 +683,110 @@ def get(
             self.columns[id] = column
             self.output_workers[id] = output_workers
             self.completed_workers[id] = set()
+            self.participating_workers[id] = output_workers.copy()
 
+        self.participating_workers[id].add(worker)
         return {
+            "status": "OK",
             "worker_for": self.worker_for[id],
             "column": self.columns[id],
             "schema": self.schemas[id],
             "output_workers": self.output_workers[id],
         }
 
+    async def remove_worker(self, scheduler: Scheduler, worker: str) -> None:
+        affected_shuffles = set()
+        broadcasts = []
+        from time import time
+
+        recs: Recs = {}
+        stimulus_id = f"shuffle-failed-worker-left-{time()}"
+        barriers = []
+        for shuffle_id, shuffle_workers in self.participating_workers.items():
+            if worker not in shuffle_workers:
+                continue
+            exception = RuntimeError(
+                f"Worker {worker} left during active shuffle {shuffle_id}"
+            )
+            self.erred_shuffles[shuffle_id] = exception
+            contact_workers = shuffle_workers.copy()
+            contact_workers.discard(worker)
+            affected_shuffles.add(shuffle_id)
+            name = self.barriers[shuffle_id]
+            barrier_task = self.scheduler.tasks.get(name)
+            if barrier_task:
+                barriers.append(barrier_task)
+                broadcasts.append(
+                    scheduler.broadcast(
+                        msg={
+                            "op": "shuffle_fail",
+                            "message": str(exception),
+                            "shuffle_id": shuffle_id,
+                        },
+                        workers=list(contact_workers),
+                    )
+                )
+
+        results = await asyncio.gather(*broadcasts, return_exceptions=True)
+        for barrier_task in barriers:
+            if barrier_task.state == "memory":
+                for dt in barrier_task.dependents:
+                    if worker not in dt.worker_restrictions:
+                        continue
```

Comment:
is this actually correct? Don't we need to reset also all other keys?

Reply:
this causes the below release do never be reached since for every task this condition is true, the scheduler already transitioned it to no-worker

Reply:
> this causes the below release do never be reached since for every task this condition is true, the scheduler already transitioned it to no-worker

Good catch, I've simplified it.

> is this actually correct? Don't we need to reset also all other keys?

All we need is to get a single task to fail. If the failing worker participated in the shuffle, there are two (possibly overlapping) scenarios:
1. The worker is an input worker: In this case, either an input task or the barrier task fail, or all input has been successfully written. Thus, we do not need to fail any task.
2. The worker is an output worker: In this case, there exists at least one output task with worker restrictions matching the worker. Transitioning this task to waiting and subsequently executing it on a different worker will cause it to fail. 

We could also reset _all_ tasks, but that would cause more work for the scheduler and possibly other workers.